crates/compiler/load_internal/src/file.rs

```diff
@@ -4370,331 +4255,117 @@ fn build_header<'a>(
         }
     };
 
-    let mut imported: Vec<(QualifiedModuleName, Vec<Loc<Ident>>, Region)> =
-        Vec::with_capacity(imports.len());
-    let mut scope_size = 0;
-
-    let mut defined_values = vec![];
-    for loc_entry in imports {
-        if let Some((qualified_module_name, exposed)) = exposed_from_import(&loc_entry.value) {
-            scope_size += num_exposes;
-
-            imported.push((qualified_module_name, exposed, loc_entry.region));
-        }
-        if let Some(value) = value_def_from_imports(arena, &filename, loc_entry)? {
-            defined_values.push(value);
-        }
-    }
-
-    let mut exposed: Vec<Symbol> = Vec::with_capacity(num_exposes);
-
-    // Make sure the module_ids has ModuleIds for all our deps,
-    // then record those ModuleIds in can_module_ids for later.
-    let mut scope: MutMap<Ident, (Symbol, Region)> =
-        HashMap::with_capacity_and_hasher(scope_size, default_hasher());
-    let home: ModuleId;
-    let name: PQModuleName;
-    let symbols_from_requires;
-
-    let ident_ids = {
+    let (name, home) = {
         // Lock just long enough to perform the minimal operations necessary.
         let mut module_ids = (*module_ids).lock();
-        let mut ident_ids_by_module = (*ident_ids_by_module).lock();
 
-        name = match opt_shorthand {
+        let name = match opt_shorthand {
             Some(shorthand) => PQModuleName::Qualified(shorthand, declared_name),
             None => PQModuleName::Unqualified(declared_name),
         };
-        home = module_ids.get_or_insert(&name);
+        let home = module_ids.get_or_insert(&name);
 
-        // Ensure this module has an entry in the ident_ids_by_module map.
-        ident_ids_by_module.get_or_insert(home);
+        (name, home)
+    };
 
-        // For each of our imports, add an entry to deps_by_name
-        //
-        // e.g. for `imports [pf.Foo.{ bar }]`, add `Foo` to deps_by_name
-        //
-        // Also build a list of imported_values_to_expose (like `bar` above.)
-        for (qualified_module_name, exposed_idents, region) in imported.into_iter() {
-            let pq_module_name = qualified_module_name.into_pq_module_name(opt_shorthand);
+    let package_entries = packages
+        .iter()
+        .map(|Loc { value: pkg, .. }| (pkg.shorthand, pkg.package_name.value))
+        .collect::<MutMap<_, _>>();
```

Comment:
is there a reason `packages` in HeaderInfo can't be in this form to begin with, instead of what it currently is?

https://github.com/roc-lang/roc/blob/e500d664fdd5b79b004ec5be4614424557b67425/crates/compiler/load_internal/src/file.rs#L4214

Just to avoid intermediate data transformations where they're not useful.

Reply:
Hm, `HeaderInfo` is just holding `PackageEntry` which comes from the parsed AST. If we transformed into `MutMap` when we crate `HeaderInfo`, we would still be doing the same number of transformations, right?

Are you suggesting that the parser should produce this `MutMap`? I think that wouldn't work well for the formatter since we'd need to preserve the order and spaces.